torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -508,12 +508,15 @@ XLAGraphExecutor::SyncTensorCollection XLAGraphExecutor::CollectSyncTensors(
           coll.indices.push_back(i);
 
           // `sharding_spec()` checks sharding equality. If IR node has no
-          // sharding, then sync XLATensor sharding to the IR node. XLATensor's
-          // sharding takes the precedence as the source of the truth.
+          // sharding, then sync XLATensor sharding to the IR node.
+          // XLATensor's sharding takes the precedence as the source of the
+          // truth.
           XLATensor::ShardingSpecPtr sharding = tensors[i]->sharding_spec();
           if (sharding) {
             dynamic_cast<XlaNode*>(ir_value.node.get())
                 ->SetSharding(sharding->sharding);
+            tensors[i]->SetIrValue(
```

Comment:
Synced offline with @JackCaoG , it's better to do this before tracing. I will revise, and see if we can do this right after execution, but before tracing.

Reply:
I vaguely remember now, before tyring out the sharded dataholder -- I ran into an issue with functionalization where the updated model param is a new tensor (without sharding), different from the input tensor (with sharding) after the computation; we can propagate the sharding annotation as we trace for the next round, so I added the data node level sharding propagation at `CollectSyncTensors` ... I think. I hope I was wrong, let's retry FYI @JackCaoG